lib/edgehog_device_forwarder/websockets/core.ex

```diff
@@ -0,0 +1,97 @@
+defmodule EdgehogDeviceForwarder.WebSockets.Core do
+  @moduledoc """
+  Module containing the simple operations for the WebSockets module.
+
+  All of its methods are meant to be used only by the WebSockets module and are meant to be run
+  inside an isolated ConCache environment.
+  """
+
+  require Logger
+
+  alias EdgehogDeviceForwarder.WebSockets.Data
+  alias EdgehogDeviceForwarderProto.Edgehog.Device.Forwarder.Message, as: ProtoMessage
+  alias EdgehogDeviceForwarderProto.Edgehog.Device.Forwarder.WebSocket, as: ProtoWebSocket
+
+  @table EdgehogDeviceForwarder.Cache.for(EdgehogDeviceForwarder.WebSockets)
+
+  @type user_socket :: pid()
+  @type message ::
+          {:text, String.t()}
+          | {:binary, binary}
+          | {:ping, binary}
+          | {:pong, binary}
+          | {:close, ProtoWebSocket.Close.t()}
+
+  @spec upgrade(EdgehogDeviceForwarder.HTTPRequests.id(), user_socket, Qex.t()) ::
+          {:websocket, user_socket}
+  def upgrade(request_id, user_socket, message_queue) do
+    for message <- message_queue do
+      send(user_socket, message)
+    end
+
+    "ws: upgrade: socket_id: #{inspect(request_id)}: created"
+    |> Logger.debug(tag: "user_event")
+
+    {:websocket, user_socket}
+  end
+
+  @spec enqueue_message(Data.socket_id(), Qex.t(), message) :: :ok
+  def enqueue_message(socket_id, message_queue, message) do
+    updated_queue = Qex.push(message_queue, message)
+    ConCache.dirty_put(@table, socket_id, {:upgrading, :websocket, updated_queue})
+  end
+
+  @spec forward(Data.socket_id(), user_socket, message) :: :ok
+  def forward(socket_id, user_socket, {:close, close_data}) do
+    close_initiate_handshake_from_device(socket_id)
+
+    # we hide our close frame structure from the socket and instead
+    # use standard socket close parameters
+    close_message = {:close, {close_data.code, close_data.reason}}
+    send(user_socket, close_message)
+
+    :ok
+  end
+
+  def forward(_socket_id, user_socket, message) do
+    send(user_socket, message)
+
+    :ok
+  end
+
+  @spec close_initiate_handshake_from_device(Data.socket_id()) :: :ok
+  def close_initiate_handshake_from_device(socket_id) do
+    "ws: forward: socket_id #{inspect(socket_id)}: close request from device"
+    |> Logger.debug(tag: "user_socket_closing")
+
+    ConCache.dirty_put(@table, socket_id, {:closing, :websocket, :device})
+  end
+
+  @spec close_initiate_handshake_from_user(Data.socket_id(), user_socket) :: :ok
+  def close_initiate_handshake_from_user(socket_id, device_socket) do
```

Comment:
See above:
```suggestion
  def user_start_socket_closing_handshake(socket_id, user_socket) do
  ```
  (note the renaming device_socket -> user_socket, too, to match the typespec)